homeassistant/components/lock/mqtt.py

```diff
@@ -57,9 +58,15 @@
     """Set up MQTT lock dynamically through MQTT discovery."""
     async def async_discover(discovery_payload):
         """Discover and add an MQTT lock."""
-        config = PLATFORM_SCHEMA(discovery_payload)
-        await _async_setup_entity(hass, config, async_add_entities,
-                                  discovery_payload[ATTR_DISCOVERY_HASH])
+        try:
+            discovery_hash = discovery_payload[ATTR_DISCOVERY_HASH]
+            config = PLATFORM_SCHEMA(discovery_payload)
+            await _async_setup_entity(hass, config, async_add_entities,
+                                      discovery_hash)
+        except Exception:
+            if discovery_hash:
+                clear_discovery_hash(hass, discovery_hash)
+            raise
```

Comment:
Where will the exception end up? Do we really want to raise?

Reply:
As discussed on discord, I don't intend to modify the error handling with these PRs, just make sure we don't end up in a broken state which can only be recovered by restarting Hass.
What you pointed to in #15446 is very interesting, but I'll leave it for another PR.